core/services/keeper/registry_synchronizer.go

```diff
@@ -0,0 +1,239 @@
+package keeper
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated/keeper_registry_wrapper"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/job"
+	"github.com/smartcontractkit/chainlink/core/services/postgres"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+const syncUpkeepQueueSize = 10
+
+func NewRegistrySynchronizer(
+	job job.Job,
+	contract *keeper_registry_wrapper.KeeperRegistry,
+	orm ORM,
+	syncInterval time.Duration,
+) *RegistrySynchronizer {
+	return &RegistrySynchronizer{
+		contract:      contract,
+		doneWG:        sync.WaitGroup{},
+		orm:           orm,
+		job:           job,
+		interval:      syncInterval,
+		chDone:        make(chan struct{}),
+		StartStopOnce: utils.StartStopOnce{},
+	}
+}
+
+// RegistrySynchronizer conforms to the job.Service interface
+var _ job.Service = &RegistrySynchronizer{}
+
+type RegistrySynchronizer struct {
+	contract *keeper_registry_wrapper.KeeperRegistry
+	doneWG   sync.WaitGroup
+	interval time.Duration
+	job      job.Job
+	orm      ORM
+
+	chDone chan struct{}
+
+	utils.StartStopOnce
+}
+
+func (rs *RegistrySynchronizer) Start() error {
+	if !rs.OkayToStart() {
+		return errors.New("RegistrySynchronizer is already started")
+	}
+	go rs.run()
+	return nil
+}
+
+func (rs *RegistrySynchronizer) Close() error {
+	if !rs.OkayToStop() {
+		return errors.New("RegistrySynchronizer is already stopped")
+	}
+	close(rs.chDone)
+	rs.doneWG.Wait()
+	return nil
+}
+
+func (rs *RegistrySynchronizer) run() {
+	rs.doneWG.Add(1)
+	ticker := time.NewTicker(rs.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-rs.chDone:
+			rs.doneWG.Done()
+			return
+		case <-ticker.C:
+			rs.syncRegistry()
+		}
+	}
+}
+
+func (rs *RegistrySynchronizer) syncRegistry() {
+	contractAddress := rs.job.KeeperSpec.ContractAddress
+	logger.Debugf("syncing registry %s", contractAddress.Hex())
+
+	err := func() error {
+		registry, err := rs.newSyncedRegistry(rs.job)
+		if err != nil {
+			return err
+		}
+		ctx, _ := postgres.DefaultQueryCtx()
+		if err = rs.orm.UpsertRegistry(ctx, &registry); err != nil {
+			return err
+		}
+		if err = rs.addNewUpkeeps(registry); err != nil {
+			return err
+		}
+		if err = rs.deleteCanceledUpkeeps(registry); err != nil {
+			return err
+		}
+		return nil
+	}()
+
+	if err != nil {
```

Comment:
A simpler way to do this with no closure would be to use `logger.ErrorIf`:
```go
var err error
defer func() { logger.ErrorIf(err, "foo bar") }()

// ...
```

Reply:
:+1: good suggestion